### PR TITLE
Add requires_grad parameter to Quaternion class

### LINKED ISSUE #3291: Quaternion with `requires_grad=False` should propagate `requires_grad=False` to results of math operations
## 🚀 Feature
This actually bundle two suggestions
- Expose a `requires_grad` kwarg in `kornia.geometry.Quaternion` constructor and class methods to allow setting `requires_grad` to `False` (can be `True` by default for backward compatibility).
- Modify the math operations dunder methods (e.g., `__add__`, `__pow__`, ...) such that when an operation is done and all operands have `requires_grad=False`, then the resulting quaternion also have `requires_grad=False`.

## Motivation

- Consistent behavior with torch tensors
- Avoid having to set `requires_grad=False` in the result of every expression

### Current behavior
```python
>>> from kornia.geometry import Quaternion
>>> q = Quaternion.identity()  # no way to set `requires_grad=False`
>>> q
Parameter containing:
tensor([1., 0., 0., 0.], requires_grad=True)
>>> q.requires_grad_(False)
Parameter containing:
tensor([1., 0., 0., 0.])
>>> q
Parameter containing:
tensor([1., 0., 0., 0.])
>>> q**2
Parameter containing:
tensor([1., 0., 0., 0.], requires_grad=True)  # expected this to have `requires_grad=False`
```
### Desired behavior
```python
>>> from kornia.geometry import Quaternion
>>> q = Quaternion.identity(requires_grad=False)  # Can set `requires_grad`
>>> q
Parameter containing:
tensor([1., 0., 0., 0.])
>>> q**2
Parameter containing:
tensor([1., 0., 0., 0.])  # Result also does not `requires_grad`
>>> q * Quaternion.identity()  # the second operand requires grad by default
Parameter containing:
tensor([1., 0., 0., 0.], requires_grad=True)  #   
```

## Alternatives

Suggested solution involves modifying dunder methods of `Quaternion`. I can implement it.

## Additional context

**EDIT**: To be honest, the fact that `Quaternion` is a subclass of `nn.Module` is a bit strange to me as: 1) it doesn't have a forward method and 2) it doesn't build a graph with the quaternion operations (as `Parameter(data)` is being called in the init, causing a detached tensor). But as I'm currently not backpropagating through quaternion operations, this is not an issue for me, but I could suggest subclassing a `torch.Tensor` instead and defining `__torch_function__` for the math operations.

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by funnym0nk3y**
This commit addresses issue #3291 by adding support for explicit gradient tracking control in the Quaternion class, making its behavior consistent with PyTorch tensors.

Changes:
- Added optional `requires_grad` parameter to Quaternion.__init__()
  - When None (default), preserves the requires_grad flag from input data
  - When specified, explicitly sets the requires_grad flag on the tensor

- Added `requires_grad` parameter to all class factory methods:
  - identity(): defaults to False
  - from_coeffs(): defaults to False
  - from_matrix(): defaults to None (inherits from matrix)
  - from_euler(): defaults to None (inherits from input angles)
  - from_axis_angle(): defaults to None (inherits from axis_angle)
  - random(): defaults to False

- Mathematical operations already preserve requires_grad correctly through PyTorch's automatic gradient propagation

The implementation ensures backward compatibility - all existing code will continue to work as the new parameter has sensible defaults.

Fixes #3291

#### Changes

Fixes # (issue)

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

### CODE REVIEW

File: kornia/geometry/quaternion.py [hunk 1]
```diff
@@ -89,12 +91,16 @@ def __init__(self, data: Union[Tensor, Parameter]) -> None:
...
+            >>> # Explicitly set requires_grad
+            >>> q3 = Quaternion(data, requires_grad=True)
 
         """
         super().__init__()
         if not isinstance(data, (Tensor, Parameter)):
             raise TypeError(f"Expected Tensor or Parameter, got {type(data)}")
         # KORNIA_CHECK_SHAPE(data, ["B", "4"])  # FIXME: resolve shape bugs. @edgarriba
+        if requires_grad is not None and not isinstance(data, Parameter):
+            data = data.requires_grad_(requires_grad)
```
Comment (Meggison) [inline kornia/geometry/quaternion.py:103]: Test if gradient is being properly now, I'm not sure if this would work for the gradient initialisation since the Quartenion class doesn't directly track gradients. 

an example of how the gradients are currently gotten:

```python
import torch
from kornia.geometry import Quaternion

q = Quaternion.identity()       # q.data is a tensor, requires_grad=False initially
q.data.requires_grad_(True)     # enable autograd on the underlying tensor

out = (q.w ** 2).sum()          # out is a torch.Tensor (scalar)
out.backward()

print("q.data:", q.data)  # returns q.data: tensor([1., 0., 0., 0.], requires_grad=True)
```

**Comment by boschet** [comment]
Should I close or is this still considered for merge?

**Comment by edgarriba** [comment]
added comments here  https://github.com/kornia/kornia/issues/3291

**Comment by edgarriba** [comment]
We are doing cleanup. Please, rebase and reopen if you still have interest in merging this PR.
